### PR TITLE
bug fixes in gradient_layer and sms_recovery

### PR DESCRIPTION
**Pull request by jediofgever**
Signed-off-by: jediofgever <fetulahatas1@gmail.com>

I was building packages in `navigation2_tutorials` and faced with a few critical bugs that didn't let me to compile `nav2_gradient_costmap_plugin` and `nav2_sms_recovery`. The causing issues were investigated and addressed in the commit below. 

A few critical corrections were made ; 
* There was a few places where `weak_ptr` was not being used properly, for instance `node_` is a weak_ptr, in needs to be locked as;

https://github.com/ros-planning/navigation2_tutorials/blob/96ddd87f194c7101ece71fc7d2ba61808549ea7f/nav2_gradient_costmap_plugin/src/gradient_layer.cpp#L73

* There was a incompetency between virtual base function and overridden function's signatures in `nav2_straightline_planner/src/straight_line_planner.cpp.  `

https://github.com/ros-planning/navigation2_tutorials/blob/96ddd87f194c7101ece71fc7d2ba61808549ea7f/nav2_straightline_planner/include/nav2_straightline_planner/straight_line_planner.hpp#L67

other changes are mostly cosmetic, running a linter, adding missing complete licences , namespace corrections and proper header guards .

### CODE REVIEW

File: nav2_sms_recovery/include/nav2_sms_recovery/sms_recovery.hpp [hunk 1]
```diff
@@ -1,8 +1,19 @@
 // Copyright (c) 2020 Samsung Research America
-// This code is licensed under MIT license (see LICENSE.txt for details)
-
-#ifndef NAV2_SMS_RECOVEY__SMS_RECOVERY_HPP_
-#define NAV2_SMS_RECOVEY__SMS_RECOVERY_HPP_
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (SteveMacenski) [inline nav2_sms_recovery/include/nav2_sms_recovery/sms_recovery.hpp:3]: You may not relicense someone elses' work.

File: nav2_sms_recovery/include/nav2_sms_recovery/type_conversion.hpp [hunk 2]
```diff
@@ -1,32 +1,60 @@
-// Taken from https://www.twilio.com/docs/sms/tutorials/how-to-send-sms-messages-cpp
-// Under MIT license
+// MIT License
+//
+// Copyright (c) [year] [fullname]
```
Comment (SteveMacenski) [inline nav2_sms_recovery/include/nav2_sms_recovery/type_conversion.hpp:3]: Please fill in here and in other headers you added

File: nav2_sms_recovery/include/nav2_sms_recovery/type_conversion.hpp [hunk 3]
```diff
@@ -1,32 +1,60 @@
...
+#ifndef NAV2_SMS_RECOVERY__TYPE_CONVERSION_HPP_
+#define NAV2_SMS_RECOVERY__TYPE_CONVERSION_HPP_
 
 #pragma once
 
 #include <locale>
 #include <codecvt>
 #include <string>
 
+namespace nav2_sms_recovery {
```
Comment (SteveMacenski) [inline nav2_sms_recovery/include/nav2_sms_recovery/type_conversion.hpp:32]: Add the namespace back in.

File: nav2_sms_recovery/src/sms_recovery.cpp [hunk 4]
```diff
@@ -21,42 +21,45 @@ SMSRecovery::~SMSRecovery()
...
+	}
+
+	RCLCPP_INFO(node->get_logger(), "SMS sent successfully!");
+	return(Status::SUCCEEDED);
 }
 
 Status SMSRecovery::onCycleUpdate()
 {
-  return Status::SUCCEEDED;
+	return(Status::SUCCEEDED);
```
Comment (SteveMacenski) [inline nav2_sms_recovery/src/sms_recovery.cpp:61]: Remove parentheses

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
Ok - we didn't lint this code because this is all demo code, but I'm cool merging it.

File: nav2_sms_recovery/src/sms_recovery.cpp [hunk 5]
```diff
@@ -4,14 +4,14 @@
...
+#include <string>
 
 #include "nav2_sms_recovery/sms_recovery.hpp"
 
 namespace nav2_sms_recovery
 {
-
 SMSRecovery::SMSRecovery()
-: Recovery<Action>()
+	: Recovery <Action>()
```
Comment (SteveMacenski) [inline nav2_sms_recovery/src/sms_recovery.cpp:14]: No, remove **all** tabs from all files. We do not allow for tabs. 2 space is google C++ styling that we follow. Revert across all files.

File: nav2_sms_recovery/src/sms_recovery.cpp [hunk 6]
```diff
@@ -21,42 +21,45 @@ SMSRecovery::~SMSRecovery()
...
+	std::string response;
+	bool message_success = _twilio->send_message(
+		_to_number,
+		_from_number,
+		command->message,
+		response,
+		"",
+		false);
+
+	if (!message_success)
```
Comment (SteveMacenski) [inline nav2_sms_recovery/src/sms_recovery.cpp:49]: This is not properly linted, you removed the open bracket from the if line, which is the C++ google styling that we follow.

**Review by SteveMacenski — COMMENTED** [review]
I'm not sure what linters you used, but you didn't use the ROS2 linters and there are a number of changes here that must be reverted.

File: nav2_sms_recovery/include/nav2_sms_recovery/sms_recovery.hpp (hunk 1 above)
Comment (jediofgever) [reply nav2_sms_recovery/include/nav2_sms_recovery/sms_recovery.hpp:3]: yes but I did not touch the copyright holder, just added the rest of standard license

File: nav2_sms_recovery/src/sms_recovery.cpp (hunk 6 above)
Comment (jediofgever) [reply nav2_sms_recovery/src/sms_recovery.cpp:49]: can you please post the `uncrustify.cfg` file that you use as a gist ? 
I have hard time to find the exact settings that is used across navigation software

**Comment by jediofgever** [comment]
I guess thats right, since this is just demo code it should be fine. I will revert the changes and only address the real issues.

File: nav2_sms_recovery/src/sms_recovery.cpp (hunk 6 above)
Comment (SteveMacenski) [reply nav2_sms_recovery/src/sms_recovery.cpp:49]: I don't have one, use `ament_uncrustify` and it's preconfigured with the proper styling. Same with ament_cpplint.

File: nav2_sms_recovery/src/sms_recovery.cpp [hunk 7]
```diff
@@ -21,19 +21,21 @@ SMSRecovery::~SMSRecovery()
...
+  node->declare_parameter("from_number");
+  _from_number = node->get_parameter("from_number").as_string();
+  node->declare_parameter("to_number");
+  _to_number = node->get_parameter("to_number").as_string();
   _twilio = std::make_shared<twilio::Twilio>(_account_sid, _auth_token);
 }
 
 Status SMSRecovery::onRun(const std::shared_ptr<const Action::Goal> command)
 {
+ 	auto node = node_.lock();
```
Comment (SteveMacenski) [inline nav2_sms_recovery/src/sms_recovery.cpp:38]: Tabs, not correctly indented

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_sms_recovery/src/sms_recovery.cpp (hunk 7 above)
Comment (SteveMacenski) [reply nav2_sms_recovery/src/sms_recovery.cpp:38]: You should change the settings on your editor to not use tabs unless your organization standardizes around it.

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**